### DISCUSSION TITLE
When waiting for long operation and token expired

### DISCUSSION THREAD

**Discussion by F6PSP** [d365collaborative/d365fo.tools#580] [Q&A]
Hi

I really enjoy the d365fo.tools Module and I really appreciate all the works that has been done.
Currently I start using the Invoke-D365LcsDatabaseRefresh and the Invoke-D365LcsDatabaseExport.
To wait for the completion of the database operation, I use the Get-D365LcsDatabaseOperationStatus with the parameter $WaitForCompletion before continuing other processes.  

My question is about on how do you handle the expiration of the Token... let's say that the operation finally take more than 3600 seconds, what is the duration of the token.  In that scenario, I should have a System.Net.WebException with status code 401 returned.

Should I add the $EnableException switch parameter and do a try/catch in my own script ?
Maybe is more a question about PSFramework on how it handles the throw of the exception, but I would like to know if someone already handle this and what you would suggest to handle that scenario...

Thx

Francis

**Comment by Splaxi** [comment]
You will have to built the logic your self.

Some pointers:

Get-D365LcsApiToken will get a fresh token object, with some details in the output.
You could pipe that directly into Set-D365LcsApiConfig (with some more details as parameters).

Armed with this, you can do some calculation on the output from Get-D365LcsApiConfig, which contains activetokenexpireson - that is a unix epoc, which you can translate into a datetime.

Or you could just pull a new token, in a loop - just before you do Get-D365LcsDatabaseOperationStatus (but without the -WaitForCompletion parameter). Get-D365LcsApiConfig | Invoke-D365LcsApiRefreshToken | Set-D365LcsApiConfig (with some more details as parameters)

Pseudo code:
```powershell
Get-D365LcsApiToken ... | Set-D365LcsApiConfig ...

$temp = Invoke-D365LcsDatabaseRefresh ...
do {
Start-Sleep -Seconds 60
Get-D365LcsApiConfig | Invoke-D365LcsApiRefreshToken | Set-D365LcsApiConfig
$status = $temp | Get-D365LcsDatabaseOperationStatus
}
while (...)
```

I know you're not the first to hit this issue - so I would encourage to write a blog post and even do a PR with a wiki (md) file, where you show how it should be done 😉 Then people can see how it should be done. More than happy to give some pointers and help make the code more solid if you like.

**Reply by F6PSP** [reply]
Thanx a lot for your reply.

It's a good proposal.  I have also check the discussion that you have in the past : 
https://github.com/d365collaborative/d365fo.tools/issues/265
About some settings that can be configured....

I saw also the way that you handle the 429 in the invoke-request handler script....
Maybe it can be handle at this very low level... 

Thx again

**Reply by Splaxi** [reply]
Sure thing, let me know if you need more details. 

If the module were to support token expiration like 429, it would require some refactoring. You are welcome to do a PR and then we can work together on it.
